python/mlx/nn/layers/positional_encoding.py

```diff
@@ -142,3 +142,47 @@ def __call__(self, x):
             y = y * self.scale
 
         return y
+
+
+class ALiBi(Module):
+    def __init__(self):
+        super().__init__()
+
+    @staticmethod
+    def create_alibi_matrix(
+        q_sequence_length: int,
+        k_sequence_length: int,
+        num_heads: int,
+        offset: int,
+        dtype=mx.float32,
+    ):
+        x1 = mx.arange(offset, q_sequence_length)
+        x2 = mx.arange(0, k_sequence_length)
+        distance_matrix = -mx.abs(
+            mx.expand_dims(x1[:, None] - x2[None, :], axis=(0, 1))
+        )
+        alibi_slope = ALiBi.create_alibi_slope(num_heads=num_heads)
+        alibi_mask = (distance_matrix * alibi_slope).astype(dtype)
+        return alibi_mask
+
+    @staticmethod
+    def create_alibi_slope(num_heads):
+        x = (2**8) ** (1 / num_heads)
+        out = mx.power(x, -mx.arange(1, num_heads + 1))
+        return mx.expand_dims(out, axis=(-1, -2))
+
+    def __call__(self, attention_scores, offset=0, mask=None):
+        q_sequence_length = attention_scores.shape[-2] + offset
+        k_sequence_length = attention_scores.shape[-1]
+        num_heads = attention_scores.shape[1]
+        dtype = attention_scores.dtype
+        alibi_mask = ALiBi.create_alibi_matrix(
+            q_sequence_length=q_sequence_length,
+            k_sequence_length=k_sequence_length,
+            num_heads=num_heads,
+            offset=offset,
+            dtype=dtype,
```

Comment:
My point was more that we don't need the local variables. They just make these lines harder to parse.

Reply:
Done, removed local variables and passed the values to the function directly.